abd-clam/src/cakes/knn/sieve_v1.rs

```diff
@@ -29,257 +27,154 @@ where
-                        .iter()
-                        .map(|&i| (i, self.tree.data().query_to_one(self.query, i)))
-                        .map(|(i, d)| (i, OrdNumber { number: d }));
-
-                self.hits.extend(new_hits);
-            });
-
-            if self.hits.len() > self.k {
-                self.trim_hits();
-            }
-
-            self.is_refined = true;
-        } else {
-            let (leaves, non_leaves): (Vec<_>, Vec<_>) = insiders
-                .into_iter()
-                .chain(straddlers.into_iter())
-                .partition(|g| g.c.is_leaf());
-
-            let children = non_leaves
-                .into_iter()
-                .flat_map(|g| {
-                    g.c.children()
-                        .unwrap_or_else(|| unreachable!("This is only called on non-leaves."))
-                })
-                .map(|c| (c, c.distance_to_instance(self.tree.data(), self.query)))
-                .map(|(c, d)| Grain::new(c, d + c.radius, c.cardinality));
-
-            self.grains = leaves.into_iter().chain(children).collect();
-        }
-    }
-
-    /// Trims hits to contain only the k-nearest neighbors.
-    pub fn trim_hits(&mut self) {
-        while self.hits.len() > self.k {
-            self.hits
-                .pop_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty and has at least k elements."));
-        }
-    }
+            .chain(insiders.into_iter())
+            .flat_map(|g| {
+                g.c.children()
+                    .unwrap_or_else(|| unreachable!("This is only called on non-leaves."))
+            })
+            .map(|c| (c, c.distance_to_instance(data, query)))
+            .map(|(c, d)| Grain::new(c, d))
+            .chain(leaves.into_iter())
+            .collect();
+    }
+
+    hits.extract()
+}
 
-    /// Returns the indices and distances to the query of the k nearest neighbors of the query.
-    pub fn extract(&self) -> Vec<(usize, U)> {
-        self.hits.iter().map(|(&i, &OrdNumber { number: d })| (i, d)).collect()
-    }
+/// Adds the instances in grains to the hits.
+#[allow(clippy::needless_for_each)]
+fn grains_to_hits<T, U, D>(data: &D, query: T, hits: &mut Hits<usize, U>, grains: &[Grain<T, U>])
+where
+    T: Send + Sync + Copy,
+    U: Number,
+    D: Dataset<T, U>,
+{
+    grains.iter().for_each(|g| {
+        let indices = g.c.indices(data);
+        let distances = data.query_to_many(query, indices);
+        hits.push_batch(indices.iter().copied().zip(distances.into_iter()));
+    });
 }
 
 /// A Grain is a structure which stores a cluster, a distance, and a multiplicity.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Copy)]
 struct Grain<'a, T: Send + Sync + Copy, U: Number> {
-    /// Just something we have to do.
-    t_: std::marker::PhantomData<T>,
     /// The cluster.
     c: &'a Cluster<T, U>,
     /// The distance of the cluster's center to the query.
     d: U,
```

Comment:
I think this documentation is incorrect because even though creating a new grain requires passing in the distance from the cluster's center to the query, that's not what d actually represents since it gets immediately set to d + radius and represents d + radius in the algorithm thereafter.